# 📊 Modül 1: Büyük Resim (EDA & Data Mart)

**Amaç:** Olist veritabanındaki 9 farklı tabloyu analiz etmek, temizlemek ve modellemeye hazır tek bir 'Master Table' (Data Mart) haline getirmek.

**Adımlar:**
1.  **Veri Bağlantısı:** PostgreSQL'den verileri çekme.
2.  **Kalite Kontrolü:** Eksik veri ve tip analizi.
3.  **Görselleştirme:** Sipariş dağılımları ve müşteri lokasyonları.
4.  **Birleştirme:** `Orders` + `Items` + `Customers` + `Products` -> `Master_Data`.

In [17]:
import sys
import os

# Docker içinden değil, lokalden bağlandığımız için host'u 'localhost' yapıyoruz
os.environ["POSTGRES_HOST"] = "localhost"

import pandas as pd
import polars as pl
import plotly.express as px
from sqlalchemy import create_engine, text

# Proje kök dizinine erişim (src modülünü bulmak için)
sys.path.append(os.path.abspath('..'))

from src.config import DATABASE_URL

print("✅ Kütüphaneler yüklendi ve konfigürasyon alındı.")

✅ Kütüphaneler yüklendi ve konfigürasyon alındı.


In [18]:
# Veritabanı Bağlantısı
engine = create_engine(DATABASE_URL)

def get_table_info():
    query = """
    SELECT table_name, 
           (SELECT COUNT(*) FROM information_schema.columns WHERE table_name=t.table_name) as columns
    FROM information_schema.tables t 
    WHERE table_schema = 'public';
    """
    with engine.connect() as conn:
        df = pd.read_sql(query, conn)
    return df

print("Veritabanındaki Tablolar:")
get_table_info()

Veritabanındaki Tablolar:


table_name  columns
0                            sellers        4
1                      order_reviews        7
2  product_category_name_translation        2
3                           products        9
4                             orders        8
5                        order_items        7
6                          customers        5
7                        geolocation        5
8                     order_payments        5

## 🚀 Adım 1: Verileri Okuma ve Birleştirme (Data Mart)

Analiz yapabilmek için parçalı tabloları (Orders, Items, Products, Customers) tek bir **Master Tablo** haline getirmeliyiz.
Bunun için `Polars` kütüphanesini kullanacağız (Pandas'tan çok daha hızlıdır).

In [19]:
# Tabloları Polars ile okuyoruz
print("⏳ Tablolar okunuyor...")

df_orders = pl.read_database("SELECT * FROM orders", connection=engine)
df_items = pl.read_database("SELECT * FROM order_items", connection=engine)
df_products = pl.read_database("SELECT * FROM products", connection=engine)
df_customers = pl.read_database("SELECT * FROM customers", connection=engine)

print(f"✅ Veriler yüklendi:\n- Orders: {df_orders.height} satır\n- Items: {df_items.height} satır\n- Products: {df_products.height} satır\n- Customers: {df_customers.height} satır")

⏳ Tablolar okunuyor...
✅ Veriler yüklendi:
- Orders: 99441 satır
- Items: 112650 satır
- Products: 32951 satır
- Customers: 99441 satır


In [20]:
# Veri Birleştirme (Join İşlemleri)
# Mantık: Order Items (En detaylı tablo) -> Orders -> Products -> Customers

master_df = (
    df_items.join(df_orders, on="order_id", how="left")
    .join(df_products, on="product_id", how="left")
    .join(df_customers, on="customer_id", how="left")
)

print(f"🧩 Master Tablo Oluşturuldu: {master_df.height} satır, {master_df.width} sütun")
master_df.head()

🧩 Master Tablo Oluşturuldu: 112650 satır, 26 sütun


shape: (5, 26)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ order_id  ┆ order_ite ┆ product_i ┆ seller_id ┆ … ┆ customer_ ┆ customer_ ┆ customer_ ┆ customer │
│ ---       ┆ m_id      ┆ d         ┆ ---       ┆   ┆ unique_id ┆ zip_code_ ┆ city      ┆ _state   │
│ str       ┆ ---       ┆ ---       ┆ str       ┆   ┆ ---       ┆ prefix    ┆ ---       ┆ ---      │
│           ┆ i64       ┆ str       ┆           ┆   ┆ str       ┆ ---       ┆ str       ┆ str      │
│           ┆           ┆           ┆           ┆   ┆           ┆ i64       ┆           ┆          │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 00010242f ┆ 1         ┆ 4244733e0 ┆ 48436dade ┆ … ┆ 871766c58 ┆ 28013     ┆ campos    ┆ RJ       │
│ e8c5a6d1b ┆           ┆ 6e7ecb497 ┆ 18ac8b2bc ┆   ┆ 55e863f6e ┆           ┆ dos goyta ┆          │
│ a2dd792cb ┆           ┆ 0a6e2683c ┆ e089ec2a0 ┆   ┆ ccc05f988 ┆           ┆ cazes     ┆          │
│ 162…      ┆           ┆ 13e…      ┆ 412…      ┆   ┆ b23…      ┆           ┆           ┆          │
│ 00018f77f ┆ 1         ┆ e5f2d52b8 ┆ dd7ddc04e ┆ … ┆ eb28e67c4 ┆ 15775     ┆ santa fe  ┆ SP       │
│ 2f0320c55 ┆           ┆ 02189ee65 ┆ 1b6c2c614 ┆   ┆ c0b838460 ┆           ┆ do sul    ┆          │
│ 7190d7a14 ┆           ┆ 8865ca93d ┆ 352b383ef ┆   ┆ 50ddfb8a3 ┆           ┆           ┆          │
│ 4bd…      ┆           ┆ 83a…      ┆ e2d…      ┆   ┆ 5d0…      ┆           ┆           ┆          │
│ 000229ec3 ┆ 1         ┆ c777355d1 ┆ 5b51032ed ┆ … ┆ 3818d81c6 ┆ 35661     ┆ para de   ┆ MG       │
│ 98224ef6c ┆           ┆ 8b72b67ab ┆ dd242adc8 ┆   ┆ 709e39d06 ┆           ┆ minas     ┆          │
│ a0657da4f ┆           ┆ beef9df44 ┆ 4c38acab8 ┆   ┆ b2738a8d3 ┆           ┆           ┆          │
│ c70…      ┆           ┆ fd0…      ┆ 8f2…      ┆   ┆ a24…      ┆           ┆           ┆          │
│ 00024acbc ┆ 1         ┆ 7634da152 ┆ 9d7a1d34a ┆ … ┆ af861d436 ┆ 12952     ┆ atibaia   ┆ SP       │
│ df0a6daa1 ┆           ┆ a4610f159 ┆ 505240900 ┆   ┆ cfc08b2c2 ┆           ┆           ┆          │
│ e931b0381 ┆           ┆ 5efa32f14 ┆ 6425275ba ┆   ┆ ddefd0ba0 ┆           ┆           ┆          │
│ 14c…      ┆           ┆ 722…      ┆ 1c2…      ┆   ┆ 746…      ┆           ┆           ┆          │
│ 00042b26c ┆ 1         ┆ ac6c36230 ┆ df560393f ┆ … ┆ 64b576fb7 ┆ 13226     ┆ varzea    ┆ SP       │
│ f59d7ce69 ┆           ┆ 68f30de03 ┆ 3a51e7455 ┆   ┆ 0d441e8f1 ┆           ┆ paulista  ┆          │
│ dfabb4e55 ┆           ┆ 045865e4e ┆ 3ab94004b ┆   ┆ b2d7d446e ┆           ┆           ┆          │
│ b4f…      ┆           ┆ 100…      ┆ a5c…      ┆   ┆ 483…      ┆           ┆           ┆          │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

## 🔍 Adım 2: EDA ve Görselleştirme

Veriyi birleştirdik. Şimdi **Plotly** kullanarak veriyi tanıyalım.
Cevap aradığımız sorular:
1.  Veride eksiklikler var mı?
2.  Siparişlerin durumu ne? (Teslim edildi, iptal vb.)
3.  Zaman içinde satışlar nasıl değişmiş?
4.  En çok hangi kategoride ürün satıyoruz?

In [21]:
# 1. Eksik Veri Analizi ve Stratejisi
null_counts = master_df.null_count()
print("Eksik Veri Sayıları:")
for col in null_counts.columns:
    count = null_counts[col][0]
    if count > 0:
        print(f"❌ {col}: {count} eksik")


Eksik Veri Sayıları:
❌ order_approved_at: 15 eksik
❌ order_delivered_carrier_date: 1194 eksik
❌ order_delivered_customer_date: 2454 eksik
❌ product_category_name: 1603 eksik
❌ product_name_lenght: 1603 eksik
❌ product_description_lenght: 1603 eksik
❌ product_photos_qty: 1603 eksik
❌ product_weight_g: 18 eksik
❌ product_length_cm: 18 eksik
❌ product_height_cm: 18 eksik
❌ product_width_cm: 18 eksik


### 🛠️ Eksik Veri Stratejisi (Karar Anı)

1.  **Teslimat Tarihleri (2454 eksik):** Bunlar henüz teslim edilmemiş veya iptal edilmiş siparişler. **Lojistik Modülü** için sadece teslim edilmiş (`delivered`) siparişleri kullanacağız, bu yüzden bu eksikler sorun değil (filtrelenecek).
2.  **Ürün Bilgileri (1603 eksik):** Kategori veya boyut bilgisi olmayan ürünler. Bunları **Growth Modülü** için "Unknown" olarak dolduracağız veya analizden çıkaracağız.
3.  **Boyutlar (18 eksik):** Çok az olduğu için bu satırları eğitim setinden çıkarabiliriz.

In [22]:
# 2. Sipariş Durumu (Bar Chart - Final Düzeltme)
status_counts = (
    master_df.group_by("order_status")
    .len()
    .sort("len", descending=True)
    .to_pandas()
)

fig = px.bar(status_counts, x='len', y='order_status', 
             orientation='h', 
             title='Sipariş Durumu Dağılımı (Logaritmik)',
             text='len',
             color='order_status')

# Logaritmik eksen ayarları
# range=[0, 5.5] -> 10^5.5 ≈ 316,000 (Yazıların sığması için yeterli alan)
# dtick=1 -> Sadece 10'un ana kuvvetlerini göster (10, 100, 1k...)
fig.update_xaxes(type="log", title="Sipariş Sayısı (Log)", range=[0, 5.5], dtick=1)
fig.update_yaxes(title="Durum")

fig.update_layout(showlegend=False, margin=dict(r=100))
fig.update_traces(textposition='outside')
fig.show()

In [23]:
# 3. Aylık Satış Trendi
monthly_sales = (
    master_df.with_columns(pl.col("order_purchase_timestamp").str.to_datetime())
    .sort("order_purchase_timestamp")
    .group_by_dynamic("order_purchase_timestamp", every="1mo")
    .agg(pl.len().alias("order_count"))
    .with_columns(pl.col("order_purchase_timestamp").dt.strftime("%Y-%m").alias("month_year"))
    .to_pandas()
)

fig = px.line(monthly_sales, x='month_year', y='order_count', 
              title='Aylık Sipariş Trendi',
              markers=True)

fig.update_xaxes(title='Ay', tickangle=-45)
fig.update_yaxes(title='Sipariş Adedi')
fig.show()

In [24]:
# 4. En Çok Satan Kategoriler
top_categories = (
    master_df.group_by("product_category_name")
    .len()
    .sort("len", descending=True)
    .head(10)
    .to_pandas()
)

max_val = top_categories['len'].max()

fig = px.bar(top_categories, x='product_category_name', y='len',
             title='En Çok Satan 10 Ürün Kategorisi',
             text_auto='.2s')

fig.update_traces(marker_color='rgb(55, 83, 109)', textposition='outside')
fig.update_layout(xaxis_title="Kategori", 
                  yaxis_title="Satış Adedi", 
                  yaxis_range=[0, max_val * 1.2], 
                  margin=dict(t=50, b=100))
fig.show()

## 🏁 Bölüm Özeti ve Sonraki Adımlar

Bu bölümde veriyi tanıdık ve temizledik:
1.  **Veri Bütünlüğü:** Siparişlerin %97'si teslim edilmiş (`delivered`). İptal oranları düşük.
2.  **Eksik Veriler:** Teslimat tarihleri eksik olanlar (henüz teslim edilmeyenler) Lojistik modülünde filtrelenecek.
3.  **Kategoriler:** 'Bed_Bath_Table' ve 'Health_Beauty' en çok satan kategoriler.

**Sıradaki Adım:** `2_logistics_engine.ipynb` ile teslimat sürelerini tahmin eden makine öğrenmesi modelini kuracağız.